{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17e2b247",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38/v0/1000G.phase3.integrated.sites_only.no_MATCHED_REV.hg38.vcf\n",
      "hg38/v0/1000G.phase3.integrated.sites_only.no_MATCHED_REV.hg38.vcf.idx\n",
      "hg38/v0/1000G_omni2.5.hg38.vcf.gz\n",
      "hg38/v0/1000G_omni2.5.hg38.vcf.gz.tbi\n",
      "hg38/v0/1000G_phase1.snps.high_confidence.hg38.vcf.gz\n",
      "hg38/v0/1000G_phase1.snps.high_confidence.hg38.vcf.gz.tbi\n",
      "hg38/v0/1000G_phase3_v4_20130502.sites.hg38.vcf\n",
      "hg38/v0/1000G_phase3_v4_20130502.sites.hg38.vcf.idx\n",
      "hg38/v0/Axiom_Exome_Plus.genotypes.all_populations.poly.hg38.vcf.gz\n",
      "hg38/v0/Axiom_Exome_Plus.genotypes.all_populations.poly.hg38.vcf.gz.tbi\n",
      "hg38/v0/CrossSpeciesContamination/ContaminantNormalizationFactors.txt\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminant/meats.dict\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminant/meats.fa\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminant/meats.fa.fai\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminant/meats.fa.img\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminant/meats.min2k.db\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesContaminationMetrics.py\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesHuman/pathseq_host.bfi\n",
      "hg38/v0/CrossSpeciesContamination/CrossSpeciesHuman/pathseq_host.fa.img\n",
      "hg38/v0/GRCh38.primary_assembly.genome.fa\n",
      "hg38/v0/GRCh38_gencode.v27.refFlat.txt\n",
      "hg38/v0/Homo_sapiens_assembly38.contam.UD\n",
      "hg38/v0/Homo_sapiens_assembly38.contam.V\n",
      "hg38/v0/Homo_sapiens_assembly38.contam.bed\n",
      "hg38/v0/Homo_sapiens_assembly38.contam.mu\n",
      "hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf\n",
      "hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx\n",
      "hg38/v0/Homo_sapiens_assembly38.dict\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.alt\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.amb\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.ann\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.bwt\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.pac\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.64.sa\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.amb\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.ann\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.bwt\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.fai\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.pac\n",
      "hg38/v0/Homo_sapiens_assembly38.fasta.sa\n",
      "hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz\n",
      "hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi\n",
      "hg38/v0/Homo_sapiens_assembly38.ref_cache.tar.gz\n",
      "hg38/v0/Homo_sapiens_assembly38.tile_db_header.vcf\n",
      "hg38/v0/Homo_sapiens_assembly38.vid\n",
      "hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz\n",
      "hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi\n",
      "hg38/v0/README\n",
      "hg38/v0/TileDB/Homo_sapiens_assembly38.tile_db_header.vcf\n",
      "hg38/v0/TileDB/Homo_sapiens_assembly38.vid\n",
      "hg38/v0/WholeGenomeShotgunContam.vcf\n",
      "hg38/v0/WholeGenomeShotgunContam.vcf.idx\n",
      "hg38/v0/autosomes-1kg-minusNA12878-ALL.vcf\n",
      "hg38/v0/gencode.v27.primary_assembly.annotation.gtf\n",
      "hg38/v0/gencode.v27.rRNA.interval_list\n",
      "hg38/v0/gencode_v27_primary.tar\n",
      "hg38/v0/gencode_v27_trans_rsem.tar.gz\n",
      "hg38/v0/genome_snp_tran.tar.gz\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.gcmask.fasta\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.gcmask.fasta.fai\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.gcprof.zip\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.gendermask.bed\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.gsparams.txt\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.interval.list\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.ploidymap.txt\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.profiles.bed\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.rdmask.bed\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.svmask.fasta\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.svmask.fasta.fai\n",
      "hg38/v0/genomestrip/Homo_sapiens_assembly38.vdjregions.bed\n",
      "hg38/v0/hapmap_3.3.hg38.vcf.gz\n",
      "hg38/v0/hapmap_3.3.hg38.vcf.gz.tbi\n",
      "hg38/v0/hg38.even.handcurated.20k.chr20_test.intervals\n",
      "hg38/v0/hg38.even.handcurated.20k.intervals\n",
      "hg38/v0/scattered_calling_intervals/temp_0001_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0002_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0003_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0004_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0005_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0006_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0007_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0008_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0010_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0011_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0012_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0013_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0014_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0015_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0016_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0017_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0018_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0021_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0022_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0029_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0030_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0031_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0033_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0034_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0035_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0036_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0040_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0041_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0042_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0043_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0044_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0045_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0046_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0047_of_50/scattered.interval_list\n",
      "hg38/v0/scattered_calling_intervals/temp_0050_of_50/scattered.interval_list\n",
      "hg38/v0/wgs_calling_regions.hg38.interval_list\n",
      "hg38/v0/wgs_calling_regions.v1.interval_list\n",
      "hg38/v0/wgs_coverage_regions.hg38.interval_list\n",
      "hg38/v0/wgs_evaluation_regions.hg38.interval_list\n",
      "hg38/v0/wgs_metrics_intervals.interval_list\n"
     ]
    }
   ],
   "source": [
    "import boto3\n",
    "from botocore import UNSIGNED\n",
    "from botocore.config import Config\n",
    "\n",
    "# Configure S3 client to use unsigned requests\n",
    "s3 = boto3.client('s3', region_name='us-east-1', config=Config(signature_version=UNSIGNED))\n",
    "\n",
    "bucket = 'broad-references'\n",
    "prefix = 'hg38'\n",
    "\n",
    "paginator = s3.get_paginator('list_objects_v2')\n",
    "for page in paginator.paginate(Bucket=bucket, Prefix=prefix):\n",
    "    for obj in page.get('Contents', []):\n",
    "        print(obj['Key'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "053211c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.5\tSO:coordinate\n",
      "@SQ\tSN:chr1\tLN:248956422\tM5:6aef897c3d6ff0c78aff06ac189178dd\tAS:38\tUR:/seq/references/Homo_sapiens_assembly38/v0/Homo_sapiens_assembly38.fasta\tSP:Homo sapiens\n",
      "@SQ\tSN:chr2\tLN:242193529\tM5:f98db672eb0993dcfdabafe2a882905c\tAS:38\tUR:/seq/references/Homo_sapiens_assembly38/v0/Homo_sapiens_assembly38.fasta\tSP:Homo sapiens\n",
      "@SQ\tSN:chr3\tLN:198295559\tM5:76635a41ea913a405ded820447d067b0\tAS:38\tUR:/seq/references/Homo_sapiens_assembly38/v0/Homo_sapiens_assembly38.fasta\tSP:Homo sapien\n"
     ]
    }
   ],
   "source": [
    "import boto3\n",
    "from botocore import UNSIGNED\n",
    "from botocore.config import Config\n",
    "\n",
    "# Configure S3 client to use unsigned requests\n",
    "s3 = boto3.client('s3', region_name='us-east-1', config=Config(signature_version=UNSIGNED))\n",
    "\n",
    "\n",
    "bucket = \"broad-references\"\n",
    "key = \"hg38/v0/wgs_calling_regions.hg38.interval_list\"\n",
    "\n",
    "# Try to get the object content\n",
    "obj = s3.get_object(Bucket=bucket, Key=key)\n",
    "content = obj['Body'].read().decode('utf-8')\n",
    "\n",
    "print(content[:500])  # preview first 500 characters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "63c05e75",
   "metadata": {},
   "outputs": [
    {
     "ename": "JSONDecodeError",
     "evalue": "Expecting value: line 1 column 1 (char 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mJSONDecodeError\u001b[39m                           Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[11]\u001b[39m\u001b[32m, line 10\u001b[39m\n\u001b[32m      7\u001b[39m s3_file = \u001b[33m\"\u001b[39m\u001b[33ms3://1000genomes/1000G_2504_high_coverage/working/20201028_3202_phased/phased-manifest_July2021.tsv\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m fs.open(s3_file, \u001b[33m'\u001b[39m\u001b[33mr\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[32m---> \u001b[39m\u001b[32m10\u001b[39m     metadata = \u001b[43mjson\u001b[49m\u001b[43m.\u001b[49m\u001b[43mload\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     12\u001b[39m \u001b[38;5;66;03m# Inspect top-level keys\u001b[39;00m\n\u001b[32m     13\u001b[39m \u001b[38;5;28mprint\u001b[39m(metadata.keys())\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/__init__.py:293\u001b[39m, in \u001b[36mload\u001b[39m\u001b[34m(fp, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[39m\n\u001b[32m    274\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mload\u001b[39m(fp, *, \u001b[38;5;28mcls\u001b[39m=\u001b[38;5;28;01mNone\u001b[39;00m, object_hook=\u001b[38;5;28;01mNone\u001b[39;00m, parse_float=\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[32m    275\u001b[39m         parse_int=\u001b[38;5;28;01mNone\u001b[39;00m, parse_constant=\u001b[38;5;28;01mNone\u001b[39;00m, object_pairs_hook=\u001b[38;5;28;01mNone\u001b[39;00m, **kw):\n\u001b[32m    276\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Deserialize ``fp`` (a ``.read()``-supporting file-like object containing\u001b[39;00m\n\u001b[32m    277\u001b[39m \u001b[33;03m    a JSON document) to a Python object.\u001b[39;00m\n\u001b[32m    278\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m    291\u001b[39m \u001b[33;03m    kwarg; otherwise ``JSONDecoder`` is used.\u001b[39;00m\n\u001b[32m    292\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m293\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mloads\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    294\u001b[39m \u001b[43m        \u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobject_hook\u001b[49m\u001b[43m=\u001b[49m\u001b[43mobject_hook\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    295\u001b[39m \u001b[43m        \u001b[49m\u001b[43mparse_float\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse_float\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparse_int\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse_int\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    296\u001b[39m \u001b[43m        \u001b[49m\u001b[43mparse_constant\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse_constant\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobject_pairs_hook\u001b[49m\u001b[43m=\u001b[49m\u001b[43mobject_pairs_hook\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/__init__.py:346\u001b[39m, in \u001b[36mloads\u001b[39m\u001b[34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[39m\n\u001b[32m    341\u001b[39m     s = s.decode(detect_encoding(s), \u001b[33m'\u001b[39m\u001b[33msurrogatepass\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m    343\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m (\u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[32m    344\u001b[39m         parse_int \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m parse_float \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[32m    345\u001b[39m         parse_constant \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_pairs_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m kw):\n\u001b[32m--> \u001b[39m\u001b[32m346\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_default_decoder\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdecode\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    347\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    348\u001b[39m     \u001b[38;5;28mcls\u001b[39m = JSONDecoder\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/decoder.py:337\u001b[39m, in \u001b[36mJSONDecoder.decode\u001b[39m\u001b[34m(self, s, _w)\u001b[39m\n\u001b[32m    332\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mdecode\u001b[39m(\u001b[38;5;28mself\u001b[39m, s, _w=WHITESPACE.match):\n\u001b[32m    333\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Return the Python representation of ``s`` (a ``str`` instance\u001b[39;00m\n\u001b[32m    334\u001b[39m \u001b[33;03m    containing a JSON document).\u001b[39;00m\n\u001b[32m    335\u001b[39m \n\u001b[32m    336\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m337\u001b[39m     obj, end = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mraw_decode\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43midx\u001b[49m\u001b[43m=\u001b[49m\u001b[43m_w\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mend\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    338\u001b[39m     end = _w(s, end).end()\n\u001b[32m    339\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m end != \u001b[38;5;28mlen\u001b[39m(s):\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/decoder.py:355\u001b[39m, in \u001b[36mJSONDecoder.raw_decode\u001b[39m\u001b[34m(self, s, idx)\u001b[39m\n\u001b[32m    353\u001b[39m     obj, end = \u001b[38;5;28mself\u001b[39m.scan_once(s, idx)\n\u001b[32m    354\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[32m--> \u001b[39m\u001b[32m355\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m JSONDecodeError(\u001b[33m\"\u001b[39m\u001b[33mExpecting value\u001b[39m\u001b[33m\"\u001b[39m, s, err.value) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m    356\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m obj, end\n",
      "\u001b[31mJSONDecodeError\u001b[39m: Expecting value: line 1 column 1 (char 0)"
     ]
    }
   ],
   "source": [
    "import s3fs\n",
    "import json\n",
    "\n",
    "# Create an S3 filesystem object with unsigned access\n",
    "fs = s3fs.S3FileSystem(anon=True)\n",
    "\n",
    "s3_file = \"s3://1000genomes/1000G_2504_high_coverage/working/20201028_3202_phased/phased-manifest_July2021.tsv\"\n",
    "\n",
    "with fs.open(s3_file, 'r') as f:\n",
    "    metadata = json.load(f)\n",
    "\n",
    "# Inspect top-level keys\n",
    "print(metadata.keys())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "43421e42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "{'ETag': '\"d41d8cd98f00b204e9800998ecf8427e\"', 'LastModified': datetime.datetime(2022, 1, 21, 20, 13, 21, tzinfo=tzutc()), 'size': 0, 'name': '1000genomes/1000G_2504_high_coverage/working/20201028_3202_phased/phased-manifest_July2021.tsv', 'type': 'file', 'StorageClass': 'STANDARD', 'VersionId': 'DN4QIQIWFLUB2kCqaagZLRXr4KFKCWDg', 'ContentType': 'inode/x-empty'}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import s3fs\n",
    "import pandas as pd\n",
    "\n",
    "fs = s3fs.S3FileSystem(anon=True)\n",
    "\n",
    "s3_file = \"s3://1000genomes/1000G_2504_high_coverage/working/20201028_3202_phased/phased-manifest_July2021.tsv\"\n",
    "\n",
    "print(fs.exists(s3_file))   # Should return True\n",
    "print(fs.info(s3_file)) \n",
    "\n",
    "with fs.open(s3_file, 'r') as f:\n",
    "    for i in range(10):\n",
    "        print(f.readline().strip())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
